File: line_chart_page.py
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException
from selenium.webdriver.common.action_chains import ActionChains
from base_page import BasePage
import time

class LineChartPage(BasePage):

    def __init__(self, driver):
        super().__init__(driver)
        self.wait = WebDriverWait(driver, 30)  # Increased wait time
        try:
            self.state_dropdown = self.wait.until(
                EC.presence_of_element_located((By.CSS_SELECTOR, '.data-filter-input'))  # Updated with actual selector
            )
            print("State dropdown found")
        except TimeoutException:
            print("State dropdown not found")
            raise

    def select_state_and_wait_for_chart(self, state):
        try:
            self.select_state(self.state_dropdown, state)
            print(f"State '{state}' selected")
            # Wait for the chart points to load after selecting the state
            self.chart_points = self.wait.until(
                EC.presence_of_all_elements_located((By.CSS_SELECTOR, '.graph-representation .points'))  # Updated with actual selector
            )
            print("Chart points found")
        except TimeoutException:
            print("Chart points not found")
            raise

    def print_text_values(self):
        try:
            action = ActionChains(self.driver)
            for point in self.chart_points:
                action.move_to_element(point).perform()
                time.sleep(1)  # Wait for the text to appear
                # Directly get the text content using the locator
                tooltip_elements = self.driver.find_elements(By.CSS_SELECTOR, 'g.hoverlayer g.hovertext text')
                for tooltip_element in tooltip_elements:
                    text_content = tooltip_element.text
                    if text_content:
                        print(text_content)
                    else:
                        print("No text found")
        except TimeoutException:
            print("Text elements not found")
            raise

File: src/main/java/com/example/LineChartPage.java
package com.example;

import org.openqa.selenium.By;
import org.openqa.selenium.TimeoutException;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.support.ui.ExpectedConditions;

import java.util.List;

public class LineChartPage extends BasePage {

    private WebElement stateDropdown;
    private List<WebElement> chartPoints;

    public LineChartPage(WebDriver driver) {
        super(driver);
        try {
            stateDropdown = wait.until(ExpectedConditions.presenceOfElementLocated(By.cssSelector(".data-filter-input")));
            System.out.println("State dropdown found");
        } catch (TimeoutException e) {
            System.out.println("State dropdown not found");
            throw e;
        }
    }

    public void selectStateAndWaitForChart(String state) {
        try {
            selectState(stateDropdown, state);
            System.out.println("State '" + state + "' selected");
            chartPoints = wait.until(ExpectedConditions.presenceOfAllElementsLocatedBy(By.cssSelector(".graph-representation .points")));
            System.out.println("Chart points found");
        } catch (TimeoutException e) {
            System.out.println("Chart points not found");
            throw e;
        }
    }

    public void printTextValues() {
        try {
            Actions action = new Actions(driver);
            for (WebElement point : chartPoints) {
                action.moveToElement(point).perform();
                Thread.sleep(1000);  // Wait for the text to appear
                List<WebElement> tooltipElements = driver.findElements(By.cssSelector("g.hoverlayer g.hovertext text"));
                for (WebElement tooltipElement : tooltipElements) {
                    String textContent = tooltipElement.getText();
                    if (!textContent.isEmpty()) {
                        System.out.println(textContent);
                    } else {
                        System.out.println("No text found");
                    }
                }
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        } catch (TimeoutException e) {
            System.out.println("Text elements not found");
            throw e;
        }
    }
}